### PR TITLE
Allow dialect to create friendly names for region arguments

### PR DESCRIPTION
**Pull request by flaub**
This supersedes PR #214

### CODE REVIEW

File: include/mlir/IR/OpImplementation.h [hunk 1]
```diff
@@ -567,7 +567,8 @@ class OpAsmParser {
   virtual ParseResult parseColonType(Type &result) = 0;
 
   /// Parse a colon followed by a type of a specific kind, e.g. a FunctionType.
-  template <typename TypeType> ParseResult parseColonType(TypeType &result) {
+  template <typename TypeType>
+  ParseResult parseColonType(TypeType &result) {
```
Comment (River707) [inline include/mlir/IR/OpImplementation.h]: This seems like an unrelated format change, likely caused by a different version of clang-format.

File: include/mlir/IR/OpImplementation.h [hunk 2]
```diff
@@ -655,6 +656,13 @@ class OpAsmDialectInterface
   /// OpAsmInterface.td#getAsmResultNames for usage details and documentation.
   virtual void getAsmResultNames(Operation *op,
                                  OpAsmSetValueNameFn setNameFn) const {}
+
+  /// Get a special name to use when printing the entry block arguments of the
+  /// region contained by an operation in this dialect.  The desired name should
+  /// be streamed into 'os'.  If the result is empty, the default name will be
+  /// used.
+  virtual void getRegionArgumentName(BlockArgument *arg,
+                                     raw_ostream &os) const {}
```
Comment (River707) [inline include/mlir/IR/OpImplementation.h]: Can we set this up with the same structure as the result names? i.e. allow groups of names. That would make things more consistent and extensible.

File: lib/IR/AsmPrinter.cpp [hunk 3]
```diff
@@ -1623,12 +1623,22 @@ void OperationPrinter::numberValuesInBlock(Block &block) {
 
   // Number the block arguments. We give entry block arguments a special name
   // 'arg'.
-  SmallString<32> specialNameBuffer(isEntryBlock ? "arg" : "");
-  llvm::raw_svector_ostream specialName(specialNameBuffer);
+  const OpAsmDialectInterface *interface = nullptr;
+  if (isEntryBlock && state) {
+    if (auto *op = block.getParentOp()) {
```
Comment (River707) [inline lib/IR/AsmPrinter.cpp]: Drop trivial braces.

File: include/mlir/IR/OpImplementation.h (hunk 2 above)
Comment (River707) [reply include/mlir/IR/OpImplementation.h]: It would also be nice to add a hook to the OpAsmOpInterface, but that requires support for default implementations which is something I need to add. So I can do that as a follow up.

File: include/mlir/IR/OpImplementation.h (hunk 1 above)
Comment (flaub) [reply include/mlir/IR/OpImplementation.h]: Which version of clang-format should I be using?

File: include/mlir/IR/OpImplementation.h (hunk 2 above)
Comment (flaub) [reply include/mlir/IR/OpImplementation.h]: It's not clear to me how to group block argument names, but I did modify the hook to be more consistent with `getAsmResultNames`

File: lib/IR/AsmPrinter.cpp [hunk 4]
```diff
@@ -1619,18 +1619,30 @@ void OperationPrinter::numberValuesInRegion(Region &region) {
...
-    if (isEntryBlock) {
-      specialNameBuffer.resize(strlen("arg"));
+  for (auto arg : block.getArguments()) {
+    SmallString<32> specialNameBuffer(isEntryBlock ? "arg" : "");
+    llvm::raw_svector_ostream specialName(specialNameBuffer);
+    if (isEntryBlock)
       specialName << nextArgumentID++;
-    }
-    setValueName(arg, specialName.str());
+    if (!valueIDs.count(arg))
```
Comment (River707) [inline lib/IR/AsmPrinter.cpp]: This check should be hoisted to the top of the loop:

for (BlockArgument *arg : block.getArguments()) {
  if (!...)
    continue;

File: test/lib/TestDialect/TestDialect.cpp [hunk 5]
```diff
@@ -41,6 +41,19 @@ struct TestOpAsmInterface : public OpAsmDialectInterface {
     if (auto asmOp = dyn_cast<AsmDialectInterfaceOp>(op))
       setNameFn(asmOp, "result");
   }
+
+  void getAsmBlockArgumentNames(Block *block,
+                                OpAsmSetValueNameFn setNameFn) const final {
+    auto op = block->getParentOp();
+    if (auto arrayAttr = op->getAttrOfType<ArrayAttr>("arg_names")) {
+      auto args = block->getArguments();
+      for (unsigned i = 0; i < std::min(arrayAttr.size(), args.size()); i++) {
```
Comment (River707) [inline test/lib/TestDialect/TestDialect.cpp]: Cache the end iterator to avoid recomputation, and prefer preincrement

e = std::min(...); i != e; ++i

File: test/lib/TestDialect/TestDialect.cpp [hunk 6]
```diff
@@ -41,6 +41,19 @@ struct TestOpAsmInterface : public OpAsmDialectInterface {
     if (auto asmOp = dyn_cast<AsmDialectInterfaceOp>(op))
       setNameFn(asmOp, "result");
   }
+
+  void getAsmBlockArgumentNames(Block *block,
+                                OpAsmSetValueNameFn setNameFn) const final {
+    auto op = block->getParentOp();
+    if (auto arrayAttr = op->getAttrOfType<ArrayAttr>("arg_names")) {
```
Comment (River707) [inline test/lib/TestDialect/TestDialect.cpp]: Prefer early exit:

auto arrayAttr = ...;
if (!arrayAttr)
  return;

File: test/lib/TestDialect/TestDialect.cpp [hunk 7]
```diff
@@ -41,6 +41,19 @@ struct TestOpAsmInterface : public OpAsmDialectInterface {
...
+      auto args = block->getArguments();
+      for (unsigned i = 0; i < std::min(arrayAttr.size(), args.size()); i++) {
+        if (auto strAttr = arrayAttr.getValue()[i].dyn_cast<StringAttr>()) {
```
Comment (River707) [inline test/lib/TestDialect/TestDialect.cpp]: Drop trivial braces.

**Comment by River707** [comment]
(Github won't let me respond to an outdated conversation, which is annoying).

>It's not clear to me how to group block argument names
What I had in mind for "grouping" is just how the interface with the 'setValueName' function works, not how the IR is actually printed. If we "grouped" block arguments, it would just mean that each argument should be prefixed with that name. Though, given that block arguments function a bit differently than operation results, it should be fine to have the behavior be slightly different. This difference should be clearly marked in a comment on the method though.

**Comment by flaub** [comment]
I don’t get why performance would be a consideration for the TestDialect.

File: test/lib/TestDialect/TestDialect.cpp (hunk 5 above)
Comment (flaub) [reply test/lib/TestDialect/TestDialect.cpp]: Why?

File: test/lib/TestDialect/TestDialect.cpp (hunk 6 above)
Comment (flaub) [reply test/lib/TestDialect/TestDialect.cpp]: Why?

**Comment by flaub** [comment]
> (Github won't let me respond to an outdated conversation, which is annoying).
> 
> > It's not clear to me how to group block argument names
> > What I had in mind for "grouping" is just how the interface with the 'setValueName' function works, not how the IR is actually printed. If we "grouped" block arguments, it would just mean that each argument should be prefixed with that name. Though, given that block arguments function a bit differently than operation results, it should be fine to have the behavior be slightly different. This difference should be clearly marked in a comment on the method though.

I haven’t found a need for this feature. Could we do this as a follow up PR?

File: include/mlir/IR/OpImplementation.h (hunk 1 above)
Comment (flaub) [reply include/mlir/IR/OpImplementation.h]: I’m happy to fix this by using the same version as you. Could you share your version?

File: test/lib/TestDialect/TestDialect.cpp (hunk 6 above)
Comment (River707) [reply test/lib/TestDialect/TestDialect.cpp]: Same as the rest.

**Comment by River707** [comment]
Coding standards apply across the entire codebase, the test dialect is no excuse to be inefficient/inconsistent.

File: test/lib/TestDialect/TestDialect.cpp (hunk 5 above)
Comment (River707) [reply test/lib/TestDialect/TestDialect.cpp]: This is part of the MLIR(LLVM) coding standards:
https://llvm.org/docs/CodingStandards.html

**Comment by flaub** [comment]
Actually, all of this hooking is one way to achieve custom names. But really, all I want is a way for having custom names. Would it make more sense for MLIR to treat names as first class entities? What if a Value has an optional name? That’s the common piece between both this new hook and the existing op result hook. I don’t know where it would best be stored, but it at least makes sense for the textual representation.

File: include/mlir/IR/OpImplementation.h (hunk 1 above)
Comment (flaub) [reply include/mlir/IR/OpImplementation.h]: This is caused by:
https://github.com/tensorflow/mlir/blob/master/.clang-format#L2

**Comment by joker-eph** [comment]
> Coding standards apply across the entire codebase, the test dialect is no excuse to be inefficient/inconsistent.

I wouldn't try to be really anchored too much on performance for the test dialect, but I agree with the "consistency" argument: it is just easier if the whole codebase is following the same idioms everywhere (and when I review code I don't necessarily realize I am reading some testing only code sometimes).

File: lib/IR/AsmPrinter.cpp [hunk 8]
```diff
@@ -1619,17 +1619,30 @@ void OperationPrinter::numberValuesInRegion(Region &region) {
...
-  SmallString<32> specialNameBuffer(isEntryBlock ? "arg" : "");
-  llvm::raw_svector_ostream specialName(specialNameBuffer);
-  for (auto *arg : block.getArguments()) {
-    if (isEntryBlock) {
-      specialNameBuffer.resize(strlen("arg"));
+  for (auto arg : block.getArguments()) {
+    if (valueIDs.count(arg))
+      continue;
+    SmallString<32> specialNameBuffer(isEntryBlock ? "arg" : "");
+    llvm::raw_svector_ostream specialName(specialNameBuffer);
```
Comment (River707) [inline lib/IR/AsmPrinter.cpp]: Seems like this loop can just be reverted modulo the `if (valueIds.count(...))` check at the beginning.

**Comment by River707** [comment]
> > (Github won't let me respond to an outdated conversation, which is annoying).
> > > It's not clear to me how to group block argument names
> > > What I had in mind for "grouping" is just how the interface with the 'setValueName' function works, not how the IR is actually printed. If we "grouped" block arguments, it would just mean that each argument should be prefixed with that name. Though, given that block arguments function a bit differently than operation results, it should be fine to have the behavior be slightly different. This difference should be clearly marked in a comment on the method though.
> 
> I haven’t found a need for this feature. Could we do this as a follow up PR?

SGTM

File: lib/IR/AsmPrinter.cpp (hunk 8 above)
Comment (flaub) [reply lib/IR/AsmPrinter.cpp]: OK, on it
